### ISSUE TITLE
Trained on COCO

### ISSUE THREAD

**Issue by FomalhautB** [lucidrains/DALLE-pytorch#47]
I trained a 6 layer 8 head model on the COCO dataset (15 epochs, 82k images, 1x Tesla V100, batch size 12). The results look as follows (conditioned on text "A car on the street"):
![image](https://user-images.githubusercontent.com/14837467/108501671-efa0c200-72b1-11eb-8f5c-5441a724ea80.png)
![image](https://user-images.githubusercontent.com/14837467/108501688-f62f3980-72b1-11eb-8459-bcf5b5664d91.png)
![image](https://user-images.githubusercontent.com/14837467/108501696-f92a2a00-72b1-11eb-81e8-af6ec374916a.png)
![image](https://user-images.githubusercontent.com/14837467/108501715-fd564780-72b1-11eb-84dc-735301e9425a.png)
![image](https://user-images.githubusercontent.com/14837467/108501726-00e9ce80-72b2-11eb-8ecc-ddb38b37780e.png)
![image](https://user-images.githubusercontent.com/14837467/108501738-0515ec00-72b2-11eb-8974-d8a2754bd99d.png)
![image](https://user-images.githubusercontent.com/14837467/108501748-07784600-72b2-11eb-8232-2f56c1b85ace.png)
![image](https://user-images.githubusercontent.com/14837467/108501755-09daa000-72b2-11eb-8c4e-0cb46bcce2f3.png)
There are more or less some structures but the description "A car on the street" does not meet well (I can only recognize a "car" in a few samples). Maybe the model and the dataset are still too small to learn the text condition.

**Comment by Thierryonre** [comment]
Maybe try "An image of a car on the street"??

**Comment by lucidrains** [comment]
Doesn't look bad at all! You need to start scaling dataset and size of the attention net! Just think of the difference between a 6 layer gpt vs a 64 layer one

**Comment by vigyanik** [comment]
Did you try the COCO notebook from @mrconter1? ([https://github.com/lucidrains/DALLE-pytorch/issues/15](url). It has a bug where it mixes up image ID with annotation ID

use
`
image_id = int(info['image_id']) 
`
instead of
`
image_id = int(info['id'])
`

**Comment by FomalhautB** [comment]
> Did you try the COCO notebook from @mrconter1? [https://github.com/lucidrains/DALLE-pytorch/issues/15](url). It has a bug where it mixes up image ID with annotation ID
> 
> use
> `image_id = int(info['image_id']) `
> instead of
> `image_id = int(info['id'])`

No, I wrote my own training script. But I may have also made the same mistake. This is my code for the COCO dataset:

```python
class COCODataset(Dataset):
    def __init__(self, annot_path, image_path, image_size=256):
        self._image_path = image_path
        
        with open(annot_path) as file:
            json_file = json.load(file)
        
        self._image_size = image_size
        self._metadata = json_file['images']
        self._captions = {entry['image_id']: entry for entry in json_file['annotations']}
    
    def __getitem__(self, index):
        metadata = self._metadata[index]
        caption = self._captions[metadata['id']]
        image_path = os.path.join(self._image_path, metadata['file_name'])
        image = Image.open(image_path).convert('RGB')
        image = self._crop_image(image)
        x = np.asarray(image) / 127.5 - 1
        return torch.Tensor(x).permute(2, 0, 1), caption['caption']
    
    def _crop_image(self, image):
        width, height = image.size
        min_length = min(width, height)
        
        # center crop
        left = (width - min_length)/2
        top = (height - min_length)/2
        right = (width + min_length)/2
        bottom = (height + min_length)/2
        image = image.crop((left, top, right, bottom))
        
        # resize
        image = image.resize((self._image_size, self._image_size))
        
        return image
    
    def __len__(self):
        return len(self._metadata)
```

Is the problem you mentioned in it?

**Comment by afiaka87** [comment]
> > Did you try the COCO notebook from @mrconter1? [https://github.com/lucidrains/DALLE-pytorch/issues/15](url). It has a bug where it mixes up image ID with annotation ID
> > use
> > `image_id = int(info['image_id']) `
> > instead of
> > `image_id = int(info['id'])`
> 
> No, I wrote my own training script. But I may have also made the same mistake. This is my code for the COCO dataset:
> 
> ```python
> class COCODataset(Dataset):
>     def __init__(self, annot_path, image_path, image_size=256):
>         self._image_path = image_path
>         
>         with open(annot_path) as file:
>             json_file = json.load(file)
>         
>         self._image_size = image_size
>         self._metadata = json_file['images']
>         self._captions = {entry['image_id']: entry for entry in json_file['annotations']} # This line here
>     
>     def __getitem__(self, index):
>         metadata = self._metadata[index]
>         caption = self._captions[metadata['id']]
>         image_path = os.path.join(self._image_path, metadata['file_name'])
>         image = Image.open(image_path).convert('RGB')
>         image = self._crop_image(image)
>         x = np.asarray(image) / 127.5 - 1
>         return torch.Tensor(x).permute(2, 0, 1), caption['caption']
>     
>     def _crop_image(self, image):
>         width, height = image.size
>         min_length = min(width, height)
>         
>         # center crop
>         left = (width - min_length)/2
>         top = (height - min_length)/2
>         right = (width + min_length)/2
>         bottom = (height + min_length)/2
>         image = image.crop((left, top, right, bottom))
>         
>         # resize
>         image = image.resize((self._image_size, self._image_size))
>         
>         return image
>     
>     def __len__(self):
>         return len(self._metadata)
> ```
> 
> Is the problem you mentioned in it?

`self._captions = {int(entry['image_id']): entry for entry in json_file['annotations']}`

Hey, I'm about to dig into COCO myself. Is the mistake referenced this one?

**Comment by GuoxingY** [comment]
Did you train the dVAE by yourself or use the pre-trained model by openai?

**Comment by afiaka87** [comment]
> Did you train the dVAE by yourself or use the pre-trained model by openai?

There is also the 1024VQGAN from the taming-transformers research available. This post is older and may use the OpenAI one - not sure.

**Comment by FomalhautB** [comment]
> > Did you train the dVAE by yourself or use the pre-trained model by openai?
> 
> There is also the 1024VQGAN from the taming-transformers research available. This post is older and may use the OpenAI one - not sure.



> Did you train the dVAE by yourself or use the pre-trained model by openai?



> > Did you train the dVAE by yourself or use the pre-trained model by openai?
> 
> There is also the 1024VQGAN from the taming-transformers research available. This post is older and may use the OpenAI one - not sure.

Well, the post is actually older than the OpenAI release, I trained the dVAE by myself. I also tested VQGAN after that, it uses significantly fewer resources and the generations look sharper and clearer.

**Comment by GuoxingY** [comment]
> > > Did you train the dVAE by yourself or use the pre-trained model by openai?
> > 
> > 
> > There is also the 1024VQGAN from the taming-transformers research available. This post is older and may use the OpenAI one - not sure.
> 
> Well, the post is actually older than the OpenAI release, I trained the dVAE by myself. I also tested VQGAN after that, it uses significantly fewer resources and the generations look sharper and clearer.

I also try to train dalle with the dVAE trained by myself, but got worse results, could you share some details about training the dVAE (or VQGAN) and the dalle. Besides, the COCO dataset you used is the 2014train?

**Comment by FomalhautB** [comment]
> > > > Did you train the dVAE by yourself or use the pre-trained model by openai?
> > > 
> > > 
> > > There is also the 1024VQGAN from the taming-transformers research available. This post is older and may use the OpenAI one - not sure.
> > 
> > 
> > Well, the post is actually older than the OpenAI release, I trained the dVAE by myself. I also tested VQGAN after that, it uses significantly fewer resources and the generations look sharper and clearer.
> 
> I also try to train dalle with the dVAE trained by myself, but got worse results, could you share some details about training the dVAE (or VQGAN) and the dalle. Besides, the COCO dataset you used is the 2014train?

I can share [my code](https://gist.github.com/FomalhautB/408298b0429454c45f138f472c838a97), it uses an old version of DALLE and it is very ugly, But you can still reference it.
About the dataset: yes, I used the 2014train.

**Comment by SeungyounShin** [comment]
This is the result from  "COCO 2014 train" , almost the same dalle structure @ 1epoch 

I notice that after 1epoch loss doesn't seem to decrease and the image is still ugly.
Can I get a better result,  with more epochs?

![caronstreet](https://user-images.githubusercontent.com/20262536/141775957-8ae4bd43-5da0-495c-81f3-c4bc8f19d0df.jpg)
